File: 112-path-sum/112-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        def compute(root,value,pre):
            if value==targetSum and root==None and pre.left==None and pre.right==None: 
                return True
            
            if root==None: return False
            
            left = compute(root.left,value+root.val,root)
            right=compute(root.right,value+root.val,root)
            return left or right
        
        if root==None:return False
        if root.left==None and root.right==None and root.val==targetSum:
            return True
        return  compute(root,0,None);

File: 112-path-sum/112-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean check(TreeNode root, int targetSum){
        
        // System.out.println("called");
        if(targetSum==0 && root.left==null && root.right==null){
            return true;
        }
        if(root==null){
            return false;
        }
        
        boolean left = root.left !=null ? check(root.left,targetSum-root.left.val)  : false;
        boolean right= root.right!=null ? check(root.right,targetSum-root.right.val) : false;
        
        return left||right;
    }
    public boolean hasPathSum(TreeNode root, int targetSum) {
        if(root==null){
            return false;
        }
        
        return check(root,targetSum-root.val);
        
    }
}